### PR TITLE
Implements support for Okta TOTP factor

### LINKED ISSUE #7: feature: Support for Okta TOTP factor
Was interested in using this utility, but noticed it does not currently support the TOTP factor (only push notifications). Our users would need TOTP support to start, maybe other factors also later.

I have a very naive implementation ready, will open a PR shortly. Happy to discuss either here or in the PR, depending on your preferences.

### PR DESCRIPTION
**Pull request by lorengordon**
Refactored the factor-specific logic into Factor-classes, and added support for a TOTP factor. Adding new factors should be reasonably easy, just subclass `Factor` and fill out the abstract methods.

```
> aws-okta-processor ...
Password:
Info: Calling https://<domain>.okta.com/api/v1/authn
Select Factor:
[ 1 ] token:software:totp (GOOGLE)
[ 2 ] push (OKTA)
[ 3 ] token:software:totp (OKTA)
Selection: 1
Token: xxxxxx
Info: Calling https://<domain>.okta.com/api/v1/authn/factors/<link>/verify
Info: Calling https://<domain>.okta.com/api/v1/sessions
Info: Calling https://<domain>.okta.com/home/amazon_aws/<app>
Role: <role>
{"AccessKeyId": ...}
```

Closes #7

### CODE REVIEW

**Comment by lorengordon** [comment]
Went ahead and re-factored the factor-specific logic into Factor-based classes. I think it makes sense... Idea is that adding new factors should only require creating a new Factor-class and filling out its abstract methods.

**Comment by tarkatronic** [comment]
Thanks for working on this @lorengordon! This looks like a great start, but I do have a couple of initial hesitations.

Currently, this changes the user experience by adding a whole new prompt. I think this should generally not be taken lightly, as it could be potentially disruptive. To minimize the effect, I think we should a) Make a major version bump, and b) Allow the factor type to be specified by a command-line switch, e.g. `--factor-type=okta-push`.

The name of the `Factor` class should probably be changed, to something like `FactorBase`, `FactorAbstract`, or something along those lines, to reflect the fact that it is no longer a concrete class.

I would lean toward putting the different factor types into an enum. This should help keep some of the code a bit more uniform. This can be kept compatible with Python 2.7 by using the `enum34` package, as long as its requirement is listed with a classifier like `enum34; python_version < 3.4`.

The `factory()` method could be made a bit more dynamic, so that the `SUPPORTED_FACTORS` dictionary doesn't have to be updated with every new factor, by using `.__subclasses__`, something like:

```python
class FactorType(enum.Enum):
    PUSH = "push"
    TOTP = "token:software:totp"

@add_metaclass(abc.ABCMeta)
class Factor(object):
    @classmethod
    def factory(cls, factor_type):
        for impl in cls.__subclasses__():
            if factor_type == impl.factor_type:
                return impl
        raise ProgrammingError("Factor type not implemented: %s" % factor_type)

class FactorPush(Factor):
    factor_type = FactorType.PUSH

class FactorTOTP(Factor):
    factor_type = FactorType.TOTP
```
None of that code is tested, so I can't guarantee it works. But it should be a good starting point!

Also, I'd love to see more tests added to cover this new functionality!

**Comment by lorengordon** [comment]
@tarkatronic Ok, used your example to update the factory implementation to lookup the factor subclass dynamically. Very cool! 👍 🆒 I didn't use the actual `enum` just to avoid the extra dependency for python 2.7. A bare class works also, but if you really want an `enum` then no problem.

Also changed the matching factor key to `<factor>:<provider>` so the prompt now looks like this:

```
Select Factor:
[ 1 ] push:okta
[ 2 ] token:software:totp:google
[ 3 ] token:software:totp:okta
```

And there is already a `--factor` option on the cli, but the required value is changing since the `provider` is necessary, e.g. `--factor push:okta`.

Agreed on tests, haven't dug in too much on your test framework. I can work on that also though...

**Comment by lorengordon** [comment]
Alright, one test added, for the totp factor. I'll work up another test case tomorrow where the AUTH_MFA_RESPONSE contains multiple supported factors, to trigger the new prompt.

**Comment by lorengordon** [comment]
@tarkatronic Ok, added another test, to check the scenario with multiple supported factors in the response. Anything else?

File: tests/core/test_okta.py [hunk 1]
```diff
@@ -211,7 +213,101 @@ def test_okta_mfa_challenge(
...
+
+    @patch('aws_okta_processor.core.prompt.input')
+    @patch('aws_okta_processor.core.okta.os.chmod')
+    @patch('aws_okta_processor.core.okta.open')
+    @patch('aws_okta_processor.core.okta.os.makedirs')
+    @patch('aws_okta_processor.core.okta.print_tty')
+    @responses.activate
+    def test_okta_mfa_push_multiple_factor_challenge(
+            self,
+            mock_print_tty,
```
Comment (tarkatronic) [inline tests/core/test_okta.py]: I noticed that there is a print leaking through here from this test; I was able to eliminate it (and cut down on the number of mocks passed to the test!) with this change:
```diff
@@ -292,11 +292,11 @@ class TestOkta(TestBase):
     @patch('aws_okta_processor.core.okta.os.chmod')
     @patch('aws_okta_processor.core.okta.open')
     @patch('aws_okta_processor.core.okta.os.makedirs')
-    @patch('aws_okta_processor.core.okta.print_tty')
+    @patch('aws_okta_processor.core.okta.print_tty', new=MagicMock())
+    @patch('aws_okta_processor.core.prompt.print_tty', new=MagicMock())
     @responses.activate
     def test_okta_mfa_push_multiple_factor_challenge(
             self,
-            mock_print_tty,
             mock_makedirs,
             mock_open,
             mock_chmod,
```
Aside from that, I think this is looking great. Test coverage looks good enough to me. Nice work!

Comment (lorengordon) [reply tests/core/test_okta.py]: Appreciate that! Had no idea how to suppress that leakage... Done now!

**Comment by tahoward** [comment]
Excellent work! Merging. Will release this afternoon. Thank you for your contribution!

**Comment by tahoward** [comment]
`v1.3.0` should be available from PYPI now :)

**Pull request merged**